airbyte-integrations/bases/debezium-v1-9-2/src/main/java/io/airbyte/integrations/debezium/internals/PostgresDebeziumStateUtil.java

```diff
@@ -0,0 +1,154 @@
+      propertiesMap.put(WorkerConfig.KEY_CONVERTER_CLASS_CONFIG, JsonConverter.class.getName());
+      propertiesMap.put(WorkerConfig.VALUE_CONVERTER_CLASS_CONFIG, JsonConverter.class.getName());
+      fileOffsetBackingStore.configure(new StandaloneConfig(propertiesMap));
+      fileOffsetBackingStore.start();
+
+      final JsonConverter keyConverter = new JsonConverter();
+      keyConverter.configure(Configuration.from(properties).subset("internal.key.converter" + ".", true).asMap(), true);
+      final JsonConverter valueConverter = new JsonConverter();
+      // Make sure that the JSON converter is configured to NOT enable schemas ...
+      final Configuration valueConverterConfig = Configuration.from(properties).edit().with("internal.value.converter" + ".schemas.enable", false)
+          .build();
+      valueConverter.configure(valueConverterConfig.subset("internal.value.converter" + ".", true).asMap(), false);
+
+      offsetStorageReader = new OffsetStorageReaderImpl(fileOffsetBackingStore, properties.getProperty("name"), keyConverter,
+          valueConverter);
+      final PostgresConnectorConfig postgresConnectorConfig = new PostgresConnectorConfig(Configuration.from(properties));
+      final PostgresCustomLoader loader = new PostgresCustomLoader(postgresConnectorConfig);
+      final OffsetReader<Partition, PostgresOffsetContext, Loader> offsetReader = new OffsetReader<>(offsetStorageReader, loader);
+      final Set<Partition> partitions = Collections.singleton(new PostgresPartition(postgresConnectorConfig.getLogicalName()));
+      final Map<Partition, PostgresOffsetContext> offsets = offsetReader.offsets(partitions);
+
+      return extractLsn(partitions, offsets, loader);
+
+    } finally {
+      LOGGER.info("Closing offsetStorageReader and fileOffsetBackingStore");
+      if (offsetStorageReader != null) {
+        offsetStorageReader.close();
+      }
+
+      if (fileOffsetBackingStore != null) {
+        fileOffsetBackingStore.stop();
+      }
+    }
+  }
+
+  private OptionalLong extractLsn(final Set<Partition> partitions,
+                                  final Map<Partition, PostgresOffsetContext> offsets,
+                                  final PostgresCustomLoader loader) {
+    boolean found = false;
+    for (final Partition partition : partitions) {
+      final PostgresOffsetContext postgresOffsetContext = offsets.get(partition);
+
+      if (postgresOffsetContext != null) {
+        found = true;
+        LOGGER.info("Found previous partition offset {}: {}", partition, postgresOffsetContext.getOffset());
+      }
+    }
+
+    if (!found) {
+      LOGGER.info("No previous offsets found");
+      return OptionalLong.empty();
+    }
+
+    final Offsets<Partition, PostgresOffsetContext> of = Offsets.of(offsets);
+    final PostgresOffsetContext previousOffset = of.getTheOnlyOffset();
+
+    final Map<String, ?> offset = previousOffset.getOffset();
+
+    if (offset.containsKey(LAST_COMMIT_LSN_KEY)) {
```

Comment:
do these different keys have different semantics? Or are they just connector-specific?

Reply:
So debezium state can differ in structure based on when the state is being generated. For instance here is the state that's generated after debezium is done with initial snapshot
```
{"last_snapshot_record":true,"lsn":23897640,"txId":505,"ts_usec":1658821300732000,"snapshot":true}
```
as you can see it doesnt have `lsn_commit` attribute, it only has `lsn` 

But here is the state when debezium switches to wal log processing
```
{"transaction_id":null,"lsn_proc":23906656,"lsn_commit":23904168,"lsn":23906656,"txId":528,"ts_usec":1658824891856537}
```
In this case you can see `lsn_commit` is present. So depending on the state structure we determine which attribute should we rely on.

Reply:
By the way, what does `lsn` in the second state mean? I cannot find any documentation about this one.